oak_functions/loader/src/server.rs

```diff
@@ -87,69 +86,58 @@ impl Policy {
-    /// `constant_response_size_bytes` in the policy.
-    fn create_error_response(&self, reason: &str) -> anyhow::Result<Response<Body>> {
-        // We include the `oak-policy-padding` in all responses. So first estimate the size of
-        // the response including this header name. The diff between this size and
-        // `self.constant_response_size_bytes` gives the required padding, that will be added as the
-        // value of the header.
-        let body = format!("Reason: {}\n", reason);
-        let estimates_size = estimate_error_response_size(body.clone())?;
-        let required_padding = self.constant_response_size_bytes as i64 - estimates_size;
-
-        if required_padding < 0 {
-            anyhow::bail!("constant_response_size_bytes specified by the policy is too small.")
-        } else {
-            let padding_bytes = vec![b'0'; required_padding as usize];
-            http::response::Builder::new()
-                .status(StatusCode::INTERNAL_SERVER_ERROR)
-                .header(POLICY_PADDING_HEADER, std::str::from_utf8(&padding_bytes)?)
-                .body(body.into())
-                .context("couldn't create response")
-        }
-    }
-
-    /// Pads the response by adding a number of unspecified characters, to make the total size of
-    /// the response equal to `self.constant_response_size_bytes`
+    /// Creates an HTTP response, with status `OK`, the input `response` as its body.
     ///
-    /// The zeros are added as the header value for header `oak-policy-padding`.
-    /// Returns an error if the size of the input `response` is larger than or equal to
-    /// `self.constant_response_size_bytes`.
-    fn pad_response(&self, response: FunctionsResponse) -> anyhow::Result<Response<Body>> {
-        let response_size = estimate_size(&response)?;
-        let required_padding = self.constant_response_size_bytes as i64 - response_size;
-        if required_padding < 0 {
-            panic!(
-                "Something is very wrong. Only call `pad_response` with responses that require padding."
-            )
+    /// To conform to the security policy, the response is padded by adding a number of unspecified
+    /// characters, to make the total size of the response equal to
+    /// `self.constant_response_size_bytes`. The padding is added as the value of the
+    /// `oak-policy-padding` header. If the size of the input `response` is larger than
+    /// `self.constant_response_size_bytes` and the response is not already indicating policy
+    /// violation, then the `response` is discarded and an HTTP response indicating the policy
+    /// violation is returned. If `self.constant_response_size_bytes` is too small even for the
+    /// policy violation response, or the response already indicates a policy violation, an error is
+    /// returned.
+    fn create_http_response(&self, response: FunctionsResponse) -> anyhow::Result<Response<Body>> {
```

Comment:
Why does this return a Result? I think it should return an actual response in all cases, since even not returning a response would be a policy violation. If there is any error, it should be returned in the response.

As a side note, this function would be a great candidate for fuzzing: pass different responses as input, and check that it does not panic, and that it always returns a response of the same exact size.

#1992